{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Guide to: insert missing residues\n",
    "\n",
    "> this is a notebook to guide the modeller template procedure, the actual computation is a remix of this\n",
    ">\n",
    "> WARNING: before running this script, make sure to setup MODELLER. I suggest to use a conda environment for this. Moreover, download a input structure to complete, in the following, the example is given by the `1QG8` structure.\n",
    ">\n",
    "> Packages required: Biopython, MODELLER\n",
    "\n",
    "Adapted from: https://salilab.org/modeller/wiki/Missing_residues\n",
    "\n",
    "Often, you will encounter files in the PDB which have missing residues. Special care must be taken in this case, as MODELLER only reads the ATOM and HETATM records, not the SEQRES records, and so will not handle missing residues automatically. **Unfortunately PDB is not reliable enough to be able to automatically rely on SEQRES**. Therefore, since I want to automate this process using the SEQRES entry, I need to add extra controls such as:\n",
    "- reading REMARK 465 for missing residues\n",
    "\n",
    "One example is PDB code 1qg8, which is missing residues 134-136 and 218-231 (see the REMARK 465 lines in the PDB file). We can use Modeller to 'fill in' these missing residues by treating the original structure (with missing residues) as a template, and building a comparative model using the full sequence extracted from the SEQRES."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "                         MODELLER 10.4, 2022/10/28, r12463\n",
      "\n",
      "     PROTEIN STRUCTURE MODELLING BY SATISFACTION OF SPATIAL RESTRAINTS\n",
      "\n",
      "\n",
      "                     Copyright(c) 1989-2022 Andrej Sali\n",
      "                            All Rights Reserved\n",
      "\n",
      "                             Written by A. Sali\n",
      "                               with help from\n",
      "              B. Webb, M.S. Madhusudhan, M-Y. Shen, G.Q. Dong,\n",
      "          M.A. Marti-Renom, N. Eswar, F. Alber, M. Topf, B. Oliva,\n",
      "             A. Fiser, R. Sanchez, B. Yerkovich, A. Badretdinov,\n",
      "                     F. Melo, J.P. Overington, E. Feyfant\n",
      "                 University of California, San Francisco, USA\n",
      "                    Rockefeller University, New York, USA\n",
      "                      Harvard University, Cambridge, USA\n",
      "                   Imperial Cancer Research Fund, London, UK\n",
      "              Birkbeck College, University of London, London, UK\n",
      "\n",
      "\n",
      "Kind, OS, HostName, Kernel, Processor: 4, Linux lammps-salicari 5.4.0-144-generic x86_64\n",
      "Date and time of compilation         : 2022/10/28 20:54:52\n",
      "MODELLER executable type             : x86_64-intel8\n",
      "Job starting time (YY/MM/DD HH:MM:SS): 2023/03/21 22:13:23\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import modeller\n",
    "\n",
    "code = '1QG8'\n",
    "input = \"./input/\"\n",
    "output = \"./output/\"\n",
    "\n",
    "# create modeller environment; starting point for all modeller work\n",
    "env = modeller.Environ()\n",
    "\n",
    "# create model and sequence file for the original structure\n",
    "# the one with missing residues \n",
    "model = modeller.Model(env, file=input + code)\n",
    "\n",
    "# Alignment object used to align sequence with missing residues\n",
    "# with the one complete\n",
    "alignment = modeller.Alignment(env)\n",
    "alignment.append_model(model, align_codes=code)\n",
    "alignment.write(file=output + code + '.seq')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we perform the alignment: from the PDB REMARKs or SEQRES records, we know the missing residues. Hence, we can make an alignment between the original 1qg8 structure (as the template), with gap characters corresponding to the missing residues, and the full sequence.\n",
    "\n",
    "First, we read the full sequence from the SEQRES using the Biopython library and wrote it as fasta file.\n",
    "The `record.id` string is used to characterise the file as `PDBID:CHAINID` (see [biopython implementation](https://github.com/biopython/biopython/blob/fadc43625b3bbb47364e3b80106034079786c25c/Bio/SeqIO/PdbIO.py#L177)).\n",
    "Hence, in the code below, there is a record for each polymer entry (to check!)\n",
    "\n",
    "Resources:\n",
    "- https://biopython.org/wiki/SeqIO\n",
    "- https://biopython.org/wiki/SeqRecord\n",
    "- https://biopython.org/docs/1.75/api/Bio.SeqIO.PdbIO.html#Bio.SeqIO.PdbIO.PdbSeqresIterator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record id 1QG8:A, chain A\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO\n",
    "for record in SeqIO.parse(input+f\"{code}.pdb\", \"pdb-seqres\"):\n",
    "    print(\"Record id %s, chain %s\" % (record.id, record.annotations[\"chain\"]))\n",
    "    record.id = record.id + \"_complete\" # suffix used to distinguish it from the one with missing residues\n",
    "    SeqIO.write(record, output+f\"{record.id}.fasta\", \"fasta\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now, we load the complete sequence into the alignment object and perform the alignment with modeller.\n",
    "All results are saved in the `.ali` file\n",
    "\n",
    "Resources:\n",
    "- https://salilab.org/modeller/10.4/manual/node296.html\n",
    "- https://salilab.org/modeller/10.4/manual/node318.html#CMD:Alignment.salign"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SALIGN_____> adding the next group to the alignment; iteration    1\n"
     ]
    }
   ],
   "source": [
    "# load the complete sequence into the previous alignment object\n",
    "alignment.append(file=output+f\"{record.id}.fasta\", alignment_format='fasta')\n",
    "alignment.salign() # perform the actual alignment\n",
    "alignment.write(file=output+f'{record.id}.ali', alignment_format='PIR')\n",
    "del alignment # to avoid running multiple times this cell"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now we prepare the `AutoModel` object that will perform the modelling.\n",
    "In particular, we define a custom class that inherits AutoModel in order to model only the section we are interested in. This speeds up the computation by roughly 3 times!\n",
    "\n",
    "Note that the selection is used to keep fixed all those residues which are not modeled\n",
    "\n",
    "Resources:\n",
    "- https://salilab.org/modeller/10.4/manual/node47.html\n",
    "- https://salilab.org/modeller/manual/node69.html\n",
    "- https://salilab.org/modeller/manual/node23.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "openf___224_> Open           $(LIB)/restyp.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/resgrp.lib\n",
      "rdresgr_266_> Number of residue groups:        2\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/sstruc.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       788841     770.353     0.752\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       789369     770.868     0.753\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/resdih.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       837969     818.329     0.799\n",
      "rdrdih__263_> Number of dihedral angle types         :        9\n",
      "              Maximal number of dihedral angle optima:        3\n",
      "              Dihedral angle names                   :  Alph Phi Psi Omeg chi1 chi2 chi3 chi4 chi5\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/radii.lib\n",
      "\n",
      "Dynamically allocated memory at   amaxlibraries [B,KiB,MiB]:       851269     831.317     0.812\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/af_mnchdef.lib\n",
      "rdwilmo_274_> Mainchain residue conformation classes:  APBLE\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch.lib\n",
      "rdclass_257_> Number of classes:        5\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch1.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch2.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch3.lib\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/xs4.mat\n",
      "rdrrwgh_268_> Number of residue types:       21\n"
     ]
    }
   ],
   "source": [
    "from modeller.automodel import AutoModel\n",
    "\n",
    "modeller.log.verbose()\n",
    "env = modeller.Environ()\n",
    "\n",
    "env.io.atom_files_directory = ['./input/']\n",
    "\n",
    "class ModelSegment(AutoModel):\n",
    "    def select_atoms(self):\n",
    "        return modeller.Selection(self.residue_range('134:A', '136:A'),\n",
    "                         self.residue_range('218:A', '231:A'))\n",
    "\n",
    "loopmodel = ModelSegment(\n",
    "    env,\n",
    "    alnfile=output+f'{record.id}.ali',\n",
    "    knowns=code, # sequence to be completed\n",
    "    sequence=f'{record.id}' # target (complete) sequence\n",
    ")\n",
    "# the following force to build only one model\n",
    "loopmodel.starting_model    = 1\n",
    "loopmodel.ending_model      = 1"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Trigger the actual run. The following are performed to build the model:\n",
    "- homology search\n",
    "- a combination of energy minimization and brief NVT simulations (see `.sch` file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "openf___224_> Open           ./output/1QG8:A_complete.ali\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       863042     842.814     0.823\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       864492     844.230     0.824\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       867392     847.062     0.827\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:       873192     852.727     0.833\n",
      "\n",
      "Dynamically allocated memory at   amaxalignment [B,KiB,MiB]:      1002400     978.906     0.956\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1003348     979.832     0.957\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1004364     980.824     0.958\n",
      "\n",
      "Read the alignment from file       : ./output/1QG8:A_complete.ali\n",
      "\n",
      "Total number of alignment positions:   255\n",
      "\n",
      "  #  Code        #_Res #_Segm PDB_code    Name\n",
      "-------------------------------------------------------------------------------\n",
      "  1       1QG8     238      1        1QG8 MOL_ID  1; MOLECULE  PROTEIN (SPORE COAT POLYSACCHARIDE BIOSYNTHESIS P SPSA); CH\n",
      "  2  1QG8:A_co     255      1\n",
      "check_a_343_> >> BEGINNING OF COMMAND\n",
      "openf___224_> Open           ./input/1QG8.pdb\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1077489    1052.235     1.028\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1077489    1052.235     1.028\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1077660    1052.402     1.028\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1078510    1053.232     1.029\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1079768    1054.461     1.030\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1081672    1056.320     1.032\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1081672    1056.320     1.032\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1081762    1056.408     1.032\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1084618    1059.197     1.034\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1084618    1059.197     1.034\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1084753    1059.329     1.035\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1089037    1063.513     1.039\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1089037    1063.513     1.039\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1089235    1063.706     1.039\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1095661    1069.981     1.045\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1095661    1069.981     1.045\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1095958    1070.271     1.045\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1105580    1079.668     1.054\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1105580    1079.668     1.054\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1106030    1080.107     1.055\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1120480    1094.219     1.069\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1120480    1094.219     1.069\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1121155    1094.878     1.069\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1142813    1116.028     1.090\n",
      "\n",
      "Dynamically allocated memory at    amaxsequence [B,KiB,MiB]:      1142813    1116.028     1.090\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1143821    1117.013     1.091\n",
      "\n",
      "Dynamically allocated memory at amaxcoordinates [B,KiB,MiB]:      1176325    1148.755     1.122\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      1175434    1147.885     1.121\n",
      "\n",
      "check_ali___> Checking the sequence-structure alignment. \n",
      "\n",
      "Implied intrachain target CA(i)-CA(i+1) distances longer than  8.0 angstroms:\n",
      "\n",
      "ALN_POS  TMPL  RID1  RID2  NAM1  NAM2     DIST\n",
      "----------------------------------------------\n",
      "END OF TABLE\n",
      "check_a_344_> << END OF COMMAND\n",
      "openf___224_> Open           ${LIB}/top_heav.lib\n",
      "read_to_681_> topology.submodel read from topology file:        3\n",
      "openf___224_> Open           ${LIB}/par.lib\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1314686    1283.873     1.254\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1316286    1285.436     1.255\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1320266    1289.322     1.259\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1322266    1291.275     1.261\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1325266    1294.205     1.264\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1329766    1298.600     1.268\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1335338    1304.041     1.273\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1338138    1306.775     1.276\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1342338    1310.877     1.280\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1348638    1317.029     1.286\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1351410    1319.736     1.289\n",
      "\n",
      "Dynamically allocated memory at  amaxparameters [B,KiB,MiB]:      1352810    1321.104     1.290\n",
      "read_pa_232_> parameters    BONDS   ANGLS  DIHEDS IMPROPS    MODE\n",
      "                              227     561     661     112       0\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1392688    1360.047     1.328\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1602114    1564.564     1.528\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1634506    1596.197     1.559\n",
      "\n",
      "getf_______W> RTF restraint not found in the atoms list:\n",
      "              residue type, indices:    13     1\n",
      "              atom names           : N     -C    CA    CD\n",
      "              atom indices         :     1     0     2     3\n",
      "\n",
      "mkilst______> segment topology constructed from sequence and RTF:\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      255   2118   2169      0         0       939\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1725158    1684.725     1.645\n",
      "\n",
      "patch_______> segment topology patched using RTF:  1 ;  PRO  ;  PROP\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      255   2118   2169   2934      3467       939\n",
      "\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1817238    1774.646     1.733\n",
      "patch_______> segment topology patched using RTF:  255 ;  GLY  ;  CTER\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      255   2119   2170   2936      3468       940\n",
      "\n",
      "genseg______> segment topology constructed from sequence and RTF:\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      255   2119   2170   2936      3468       940\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      1838842    1795.744     1.754\n",
      "patch_______> segment topology patched using RTF:  154 242 ;  CYS  CYS  ;  DISU\n",
      "              segments residues  atoms  bonds angles dihedrals impropers:\n",
      "                     1      255   2119   2171   2938      3471       940\n",
      "patch_s_522_> Number of disulfides patched in MODEL:        1\n",
      "transfe_506_> MODEL is an average of all templates.\n",
      "transfe_511_> Number of templates for coordinate transfer:        1\n",
      "              After transferring coordinates of the equivalent template atoms,\n",
      "              there are defined, undefined atoms in MODEL:     1968      151\n",
      "openf___224_> Open           1QG8A_complete.ini\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      255     2119     2119\n",
      "make_re_417_> Restraint type to be calculated:  stereo\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1946291    1900.675     1.856\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1954483    1908.675     1.864\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      1970867    1924.675     1.880\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2003635    1956.675     1.911\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2036403    1988.675     1.942\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2101939    2052.675     2.005\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2134707    2084.675     2.036\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2167475    2116.675     2.067\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2233011    2180.675     2.130\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2364083    2308.675     2.255\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2429619    2372.675     2.317\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2495155    2436.675     2.380\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      2626227    2564.675     2.505\n",
      "r_stere_606_> Stereochemical restraints were constructed from RTF & PRMF.\n",
      "              Added bond,angle,dihedral,improper restraints  :     2171    2938    3293     930\n",
      "              Total number of restraints before, now         :        0     9332\n",
      "make_re_422_> Number of previous, current restraints         :        0     9332\n",
      "make_re_423_> Number of previous, current selected restraints:        0     9332\n",
      "make_re_417_> Restraint type to be calculated:  phi-psi_binormal\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      2776103    2711.038     2.647\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch1.bin\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/mnch1.mdt\n",
      "getdata_643_> Protein accepted:  1QG8\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3038263    2967.054     2.898\n",
      "make_re_422_> Number of previous, current restraints         :     9332     9585\n",
      "make_re_423_> Number of previous, current selected restraints:     9332     9585\n",
      "make_re_417_> Restraint type to be calculated:  omega_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      3038263    2967.054     2.898\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/omega.bin\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/omega.mdt\n",
      "getdata_643_> Protein accepted:  1QG8\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "omgdel__425_> Unselected all O C +N +CA dihedrals:      264\n",
      "              (This is to avoid clashes between STEREO\n",
      "              and OMEGA_DIHEDRAL restraints)\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3169335    3095.054     3.023\n",
      "make_re_422_> Number of previous, current restraints         :     9585     9839\n",
      "make_re_423_> Number of previous, current selected restraints:     9585     9575\n",
      "make_re_417_> Restraint type to be calculated:  chi1_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      3169335    3095.054     3.023\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi1.mdt\n",
      "getdata_643_> Protein accepted:  1QG8\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "make_re_422_> Number of previous, current restraints         :     9839    10074\n",
      "make_re_423_> Number of previous, current selected restraints:     9575     9810\n",
      "make_re_417_> Restraint type to be calculated:  chi2_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      3169335    3095.054     3.023\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi2.mdt\n",
      "getdata_643_> Protein accepted:  1QG8\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      3300407    3223.054     3.148\n",
      "make_re_422_> Number of previous, current restraints         :    10074    10259\n",
      "make_re_423_> Number of previous, current selected restraints:     9810     9995\n",
      "make_re_417_> Restraint type to be calculated:  chi3_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      3300407    3223.054     3.148\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi3.mdt\n",
      "getdata_643_> Protein accepted:  1QG8\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "make_re_422_> Number of previous, current restraints         :    10259    10338\n",
      "make_re_423_> Number of previous, current selected restraints:     9995    10074\n",
      "make_re_417_> Restraint type to be calculated:  chi4_dihedral\n",
      "\n",
      "Dynamically allocated memory at   amaxstructure [B,KiB,MiB]:      3300407    3223.054     3.148\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi1234.bin\n",
      "openf___224_> Open           ${MODINSTALL10v4}/modlib/chi4.mdt\n",
      "mdtrsr__446W> A potential that relies on one protein is used, yet you have at\n",
      "              least one known structure available. MDT, not library, potential is used.\n",
      "getdata_643_> Protein accepted:  1QG8\n",
      "getdata_289_> Proteins (all/accepted):        1        1\n",
      "make_re_422_> Number of previous, current restraints         :    10338    10370\n",
      "make_re_423_> Number of previous, current selected restraints:    10074    10106\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "\n",
      "Dynamically allocated memory at amaxhash_contac [B,KiB,MiB]:      3910643    3818.987     3.729\n",
      "make_re_422_> Number of previous, current restraints         :    10370    15114\n",
      "make_re_423_> Number of previous, current selected restraints:    10106    14850\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      4172787    4074.987     3.979\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      4697075    4586.987     4.479\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      4959219    4842.987     4.729\n",
      "make_re_422_> Number of previous, current restraints         :    15114    20001\n",
      "make_re_423_> Number of previous, current selected restraints:    14850    19737\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    20001    24029\n",
      "make_re_423_> Number of previous, current selected restraints:    19737    23765\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      5221363    5098.987     4.979\n",
      "make_re_422_> Number of previous, current restraints         :    24029    27100\n",
      "make_re_423_> Number of previous, current selected restraints:    23765    26836\n",
      "0 atoms in HETATM/BLK residues constrained\n",
      "to protein atoms within 2.30 angstroms\n",
      "and protein CA atoms within 10.00 angstroms\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    27100    27100\n",
      "make_re_423_> Number of previous, current selected restraints:    26836    26836\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    27100    27100\n",
      "make_re_423_> Number of previous, current selected restraints:    26836    26836\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    27100    27100\n",
      "make_re_423_> Number of previous, current selected restraints:    26836    26836\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    27100    27100\n",
      "make_re_423_> Number of previous, current selected restraints:    26836    26836\n",
      "0 atoms in residues without defined topology\n",
      "constrained to be rigid bodies\n",
      "make_re_417_> Restraint type to be calculated:  DISTANCE\n",
      "make_re_422_> Number of previous, current restraints         :    27100    27100\n",
      "make_re_423_> Number of previous, current selected restraints:    26836    26836\n",
      "rmdupl__427_>     1595 redundant cosine dihedral restraints were unselected.\n",
      "condens_443_> Restraints marked for deletion were removed.\n",
      "              Total number of restraints before, now:    27100    25241\n",
      "openf___224_> Open           1QG8A_complete.rsr\n",
      "openf___224_> Open           1QG8A_complete.rsr\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      5279984    5156.234     5.035\n",
      "rdcsr2__307_> Number of restraints read    :    25241\n",
      "              Number of excluded pairs read:        0\n",
      "              Number of pseudo atoms read  :        0\n",
      "rdcsrs__304_> Restraints in memory, selected restraints:    25241    25241\n",
      "              Explicitly excluded atom pairs in memory :        0\n",
      "              Pseudo atoms in memory                   :        0\n",
      "151 (of 2119 total) atoms selected for optimization\n",
      "openf___224_> Open           1QG8A_complete.ini\n",
      "\n",
      "Dynamically allocated memory at       amaxmodel [B,KiB,MiB]:      5280016    5156.266     5.035\n",
      "read_mo_297_> Segments, residues, atoms:        1      255     2119\n",
      "read_mo_298_> Segment:        1  1:A  255:A     2119\n",
      "randomi_498_> Atoms,selected atoms,random_seed,amplitude:     2119      151        1        4.0000\n",
      "randomi_496_> Amplitude is > 0; randomization is done.\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      650\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      674\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      685\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      689\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      689\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      689\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      689\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      695\n",
      "\n",
      "Dynamically allocated memory at  amaxrestraints [B,KiB,MiB]:      5623444    5491.645     5.363\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      695\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      724\n",
      "pick_re_612_> Number of MODEL atoms, selected restraints;      151      724\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:      -90.2469       15  15\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_484W> Dihedral still outside +-90:       90.9683\n",
      "iupac_m_486_> OE1/2 will be swapped:      179.9657      133  133\n",
      "iupac_m_487_> NH1/2 swapped:      179.2326      135  135\n",
      "iupac_m_483_> CD1/2 CE1/2 swapped:       91.1865      204  204\n",
      "iupac_m_397W> Atoms were not swapped because of the uncertainty of how to handle the H atom.\n",
      "iupac_m_484W> Dihedral still outside +-90:      -92.8380\n",
      "iupac_m_485_> OD1/2 will be swapped:      179.9802      217  217\n",
      "iupac_m_486_> OE1/2 will be swapped:     -125.7672      231  231\n",
      "\n",
      "\n",
      ">> ENERGY; Differences between the model's features and restraints:\n",
      "Number of all residues in MODEL                   :      255\n",
      "Number of all, selected real atoms                :     2119     151\n",
      "Number of all, selected pseudo atoms              :        0       0\n",
      "Number of all static, selected restraints         :    25241     724\n",
      "COVALENT_CYS                                      :        F\n",
      "NONBONDED_SEL_ATOMS                               :        1\n",
      "Number of non-bonded pairs (excluding 1-2,1-3,1-4):      230\n",
      "Dynamic pairs routine                             : 5, NATM_SEL x NATM double loop\n",
      "Atomic shift for contacts update (UPDATE_DYNAMIC) :    0.390\n",
      "LENNARD_JONES_SWITCH                              :    6.500   7.500\n",
      "COULOMB_JONES_SWITCH                              :    6.500   7.500\n",
      "RESIDUE_SPAN_RANGE                                :        0   99999\n",
      "NLOGN_USE                                         :       15\n",
      "CONTACT_SHELL                                     :    4.000\n",
      "DYNAMIC_PAIRS,_SPHERE,_COULOMB,_LENNARD,_MODELLER :        T       T       F       F       F\n",
      "SPHERE_STDV                                       :    0.050\n",
      "RADII_FACTOR                                      :    0.820\n",
      "Current energy                                    :         181.2788\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Summary of the restraint violations: \n",
      "\n",
      "   NUM     ... number of restraints.\n",
      "   NUMVI   ... number of restraints with RVIOL > VIOL_REPORT_CUT[i].\n",
      "   RVIOL   ... relative difference from the best value.\n",
      "   NUMVP   ... number of restraints with -Ln(pdf) > VIOL_REPORT_CUT2[i].\n",
      "   RMS_1   ... RMS(feature, minimally_violated_basis_restraint, NUMB).\n",
      "   RMS_2   ... RMS(feature, best_value, NUMB).\n",
      "   MOL.PDF ... scaled contribution to -Ln(Molecular pdf).\n",
      "\n",
      " #                     RESTRAINT_GROUP      NUM   NUMVI  NUMVP   RMS_1   RMS_2         MOL.PDF     S_i\n",
      "------------------------------------------------------------------------------------------------------\n",
      " 1 Bond length potential              :     156       0      0   0.009   0.009      3.6970       1.000\n",
      " 2 Bond angle potential               :     212       2      4   3.561   3.561      51.973       1.000\n",
      " 3 Stereochemical cosine torsion poten:      94       0      5  50.069  50.069      39.780       1.000\n",
      " 4 Stereochemical improper torsion pot:      70       1      1   1.996   1.996      9.4063       1.000\n",
      " 5 Soft-sphere overlap restraints     :     230       0      0   0.009   0.009      2.0133       1.000\n",
      " 6 Lennard-Jones 6-12 potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 7 Coulomb point-point electrostatic p:       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 8 H-bonding potential                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      " 9 Distance restraints 1 (CA-CA)      :      24       0      0   0.387   0.387     0.32628       1.000\n",
      "10 Distance restraints 2 (N-O)        :      34       0      0   0.768   0.768      1.2170       1.000\n",
      "11 Mainchain Phi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "12 Mainchain Psi dihedral restraints  :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "13 Mainchain Omega dihedral restraints:      19       0      0   3.999   3.999      3.5834       1.000\n",
      "14 Sidechain Chi_1 dihedral restraints:      17       0      0  78.992  78.992      3.2895       1.000\n",
      "15 Sidechain Chi_2 dihedral restraints:      16       0      0  96.191  96.191      6.2468       1.000\n",
      "16 Sidechain Chi_3 dihedral restraints:       7       0      0 129.407 129.407      7.2019       1.000\n",
      "17 Sidechain Chi_4 dihedral restraints:       2       0      0  76.036  76.036     0.98161       1.000\n",
      "18 Disulfide distance restraints      :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "19 Disulfide angle restraints         :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "20 Disulfide dihedral angle restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "21 Lower bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "22 Upper bound distance restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "23 Distance restraints 3 (SDCH-MNCH)  :      30       0      0   0.211   0.211     0.23335E-01   1.000\n",
      "24 Sidechain Chi_5 dihedral restraints:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "25 Phi/Psi pair of dihedral restraints:      21      15      7  58.824 149.862      50.511       1.000\n",
      "26 Distance restraints 4 (SDCH-SDCH)  :      22       0      0   0.823   0.823      1.0277       1.000\n",
      "27 Distance restraints 5 (X-Y)        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "28 NMR distance restraints 6 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "29 NMR distance restraints 7 (X-Y)    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "30 Minimal distance restraints        :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "31 Non-bonded restraints              :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "32 Atomic accessibility restraints    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "33 Atomic density restraints          :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "34 Absolute position restraints       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "35 Dihedral angle difference restraint:       0       0      0   0.000   0.000      0.0000       1.000\n",
      "36 GBSA implicit solvent potential    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "37 EM density fitting potential       :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "38 SAXS restraints                    :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "39 Symmetry restraints                :       0       0      0   0.000   0.000      0.0000       1.000\n",
      "\n",
      "\n",
      "\n",
      "openf___224_> Open           1QG8A_complete.V99990001\n",
      "# Heavy relative violation of each residue is written to: 1QG8A_complete.V99990001\n",
      "# The profile is NOT normalized by the number of restraints.\n",
      "# The profiles are smoothed over a window of residues:    1\n",
      "# The sum of all numbers in the file:    4486.7754\n",
      "\n",
      "\n",
      "\n",
      "List of the violated restraints:\n",
      "   A restraint is violated when the relative difference\n",
      "   from the best value (RVIOL) is larger than CUTOFF.\n",
      "\n",
      "   ICSR   ... index of a restraint in the current set.\n",
      "   RESNO  ... residue numbers of the first two atoms.\n",
      "   ATM    ... IUPAC atom names of the first two atoms.\n",
      "   FEAT   ... the value of the feature in the model.\n",
      "   restr  ... the mean of the basis restraint with the smallest\n",
      "              difference from the model (local minimum).\n",
      "   viol   ... difference from the local minimum.\n",
      "   rviol  ... relative difference from the local minimum.\n",
      "   RESTR  ... the best value (global minimum).\n",
      "   VIOL   ... difference from the best value.\n",
      "   RVIOL  ... relative difference from the best value.\n",
      "\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature  2                           : Bond angle potential                    \n",
      "List of the RVIOL violations larger than   :       4.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   3662 133E 134N C   N    1082 1084  155.69  120.00   35.69    8.11  120.00   35.69    8.11\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature  4                           : Stereochemical improper torsion potentia\n",
      "List of the RVIOL violations larger than   :       4.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   7008 133E 133E C   CA   1082 1076   13.98    0.00   13.98    4.92    0.00   13.98    4.92\n",
      "\n",
      "-------------------------------------------------------------------------------------------------\n",
      "\n",
      "Feature 25                           : Phi/Psi pair of dihedral restraints     \n",
      "List of the RVIOL violations larger than   :       6.5000\n",
      "\n",
      "    #   ICSR  RESNO1/2 ATM1/2   INDATM1/2    FEAT   restr    viol   rviol   RESTR    VIOL   RVIOL\n",
      "    1   7605 132N 133E C   N    1073 1075  179.96 -117.80   63.69    2.03  -63.60 -154.45   33.72\n",
      "    1        133E 133E N   CA   1075 1076  150.31  136.80                  -40.30\n",
      "    2   7606 133E 134N C   N    1082 1084 -163.13  -63.20  101.11   14.56  -63.20  101.11   14.56\n",
      "    2        134N 134N N   CA   1084 1085  -25.70  -41.10                  -41.10\n",
      "    3   7607 134N 135R C   N    1090 1092 -152.38 -125.20   40.07    1.24  -63.00  173.63   29.44\n",
      "    3        135R 135R N   CA   1092 1093  170.04  140.60                  -41.10\n",
      "    4   7689 216T 217D C   N    1788 1790  179.97  -96.50  100.87    4.19  -63.30 -170.52   30.69\n",
      "    4        217D 217D N   CA   1790 1791  170.75  114.20                  -40.00\n",
      "    5   7690 217D 218Q C   N    1796 1798 -164.55 -121.10   49.22    1.52  -63.80 -173.57   32.51\n",
      "    5        218Q 218Q N   CA   1798 1799  162.83  139.70                  -40.30\n",
      "    6   7691 218Q 219S C   N    1805 1807   78.63 -136.60  147.95    5.13  -64.10 -157.73   16.35\n",
      "    6        219S 219S N   CA   1807 1808 -178.33  151.20                  -35.00\n",
      "    7   7693 220I 221H C   N    1819 1821  -94.89  -67.60   34.74    3.10  -63.20  163.91   18.72\n",
      "    7        221H 221H N   CA   1821 1822  118.51  140.00                  -42.30\n",
      "    8   7694 221H 222F C   N    1829 1831 -138.45 -124.20   14.42    0.50  -63.20 -173.89   30.74\n",
      "    8        222F 222F N   CA   1831 1832  145.48  143.30                  -44.30\n",
      "    9   7695 222F 223Q C   N    1840 1842 -133.78 -121.10   18.59    0.61  -63.80 -179.47   30.23\n",
      "    9        223Q 223Q N   CA   1842 1843  153.29  139.70                  -40.30\n",
      "   10   7696 223Q 224L C   N    1849 1851 -118.39 -108.50   45.66    2.35  -63.50  151.98   24.50\n",
      "   10        224L 224L N   CA   1851 1852  177.07  132.50                  -41.20\n",
      "   11   7697 224L 225F C   N    1857 1859  -52.73  -71.40   25.00    1.43  -63.20  168.70   24.13\n",
      "   11        225F 225F N   CA   1859 1860  124.08  140.70                  -44.30\n",
      "   12   7698 225F 226E C   N    1868 1870  179.77 -117.80   73.94    2.17  -63.60 -175.25   30.82\n",
      "   12        226E 226E N   CA   1870 1871  176.41  136.80                  -40.30\n",
      "   13   7699 226E 227L C   N    1877 1879  -91.74 -108.50   36.33    1.75  -63.50  144.27   18.61\n",
      "   13        227L 227L N   CA   1879 1880  100.27  132.50                  -41.20\n",
      "   14   7701 228E 229K C   N    1894 1896  -51.10  -70.20   19.20    1.36  -62.90  179.66   23.99\n",
      "   14        229K 229K N   CA   1896 1897  138.47  140.40                  -40.80\n",
      "   15   7702 229K 230N C   N    1903 1905  -95.25  -71.20   24.23    1.68  -63.20  176.08   23.83\n",
      "   15        230N 230N N   CA   1905 1906  145.76  142.80                  -41.10\n",
      "\n",
      "\n",
      "report______> Distribution of short non-bonded contacts:\n",
      "\n",
      "  serious non-bonded atom clash:  1027 1046   2.156\n",
      "  serious non-bonded atom clash:  1072 1075   1.917\n",
      "\n",
      "DISTANCE1:  0.00 2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40\n",
      "DISTANCE2:  2.10 2.20 2.30 2.40 2.50 2.60 2.70 2.80 2.90 3.00 3.10 3.20 3.30 3.40 3.50\n",
      "FREQUENCY:     1    1    0    1    3    5   12   34   96  138  123  152  233  302  354\n",
      "\n",
      "\n",
      "<< end of ENERGY.\n",
      "openf___224_> Open           1QG8A_complete.B99990001.pdb\n",
      "wrpdb___568_> Residues, atoms, selected atoms:      255     2119     2119\n",
      "\n",
      ">> Summary of successfully produced models:\n",
      "Filename                          molpdf\n",
      "----------------------------------------\n",
      "1QG8A_complete.B99990001.pdb      181.27881\n",
      "\n"
     ]
    }
   ],
   "source": [
    "loopmodel.make()\n",
    "\n",
    "# move the output file to the output directory\n",
    "import shutil\n",
    "import os\n",
    "source = \"./\"\n",
    "files_list = [file for file in os.listdir(source) if file.startswith(code)]\n",
    "for files in files_list:\n",
    "    shutil.move(files, output)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As outputs, modeller dump various files that describes the minimization procedure as well as the final modelled structure.\n",
    "A bried recap, also if the run was succesful. are stored in the `outputs` variable of the `AutoModel` object\n",
    "\n",
    "Resources:\n",
    "- https://salilab.org/modeller/10.4/manual/node31.html#SECTION:model-outputs\n",
    "- https://salilab.org/modeller/manual/node267.html#CLASS:physvalues\n",
    "- https://salilab.org/modeller/methenz/andras/node15.html"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Success!\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'./output/1QG8A_fixed.pdb'"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# check if run was successful\n",
    "if loopmodel.outputs[0][\"failure\"] is None:\n",
    "    print(\"Success!\")\n",
    "# save output file name and rename it\n",
    "final_output_file = loopmodel.outputs[0][\"name\"]\n",
    "shutil.copyfile(output+final_output_file, output+f\"{final_output_file[:5]}_fixed.pdb\")"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Addendum\n",
    "\n",
    "From the [FAQs](https://salilab.org/modeller/FAQ.html#14)\n",
    "\n",
    "**How do I prevent “knots” in the final models?**\n",
    "\n",
    "The best way to prevent knots is to start with a starting structure that is as close to the desired final model as possible. Other than that, the only solution at this point is to calculate independently many models and hope that in some runs there won't be knots. Knots usually occur when one or more neighboring long insertions (i.e., longer than 15 residues) are modeled from scratch. The reason is that an insertion is built from a randomized distorted structure that is located approximately between the two anchoring regions. Under such conditions, it is easy for the optimizer to “fall” into a knot and then not be able to recover from it. Sometimes knots result from an incorrect alignment, especially when more than one template is used. When the alignment is correct, knots are a result of optimization not being good enough. However, making optimization more thorough by increasing the CPU time would not be worth it on the average as knots occur relatively infrequently. The excluded volume restraints are already included in standard comparative modeling with the AutoModel class"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "bio",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
